zebROS_ws/src/behaviors/src/2023_fourber_server.cpp

```diff
@@ -0,0 +1,383 @@
+    safety_high_distance_ = res;
+    load_param_helper(nh_, "safety_high/below", res_bool, false);
+    safety_high_below_ = res_bool;
+
+    load_param_helper(nh_, "safety_intake/distance", res, 0.4);
+    safety_low_distance_ = res;
+    load_param_helper(nh_, "safety_intake/below", res_bool, true);
+    safety_low_below_ = res_bool;
+
+    ddr_.publishServicesTopics();
+    as_.start();
+    FourberINFO("Started Fourber Action server");
+  }
+
+  ~FourberAction2023(void)
+  {
+  }
+
+  // blocks until fourbar is at the correct position
+  bool waitForFourbar(double position) {
+    ros::Rate r = ros::Rate(10);
+    
+    while (true) {
+      ros::spinOnce();
+      FourberINFO("Waiting for fourbar");
+
+      // essentially just keep fourbar where it is now
+			if (as_.isPreemptRequested() || !ros::ok()) {
+        FourberWARN("Fourber Preemepted");
+        controllers_2023_msgs::FourBarSrv last_req;
+        last_req.request.position = fourbar_cur_setpoint_;
+        fourbar_srv_.call(last_req);
+        return false;
+			}
+
+      if (fabs(fourbar_cur_setpoint_-position) <= position_tolerance_) {
+        FourberINFO("Elevator reached position!");
+        break;
+      }
+      r.sleep();
+    }
+
+    return true;
+  }
+
+  // min distance is the minimum distance the forbar must be extended to not cause problems
+  void safetyBoundsAndCallService(double min_distance, bool below) {
+    controllers_2023_msgs::FourBarSrv safety_req;
+
+    // wanting to go to safe position and already at a safe position
+    if (fourbar_cur_setpoint_ >= min_distance && fourbar_cur_position_ >= min_distance) {
+      // nothing to do
+      return; 
+    }
+    // we are trying to go to an illegal position, 
+    else if (fourbar_cur_setpoint_ < min_distance) {
+      safety_req.request.position = min_distance;
+    }
+    // currentlly in illegal position but want to go to legal position
+    else if (fourbar_cur_setpoint_ >= min_distance && fourbar_cur_position_ < min_distance) {
+      safety_req.request.position = fourbar_cur_setpoint_;
+    }
+
+    safety_req.request.below = below;
+    fourbar_srv_.call(safety_req);
+    if (!waitForFourbar(safety_req.request.position)) {
+      FourberERR("Failed calling fourber with message 'safety_req'");
+      publishFailure();
+    }
+    else {
+      publishSuccess();
+    }
+  }
+
+  void publishFailure() {
+    feedback_.success = false;
+    result_.success = false;
+    as_.publishFeedback(feedback_);
+    as_.setAborted(result_);
+  }
+
+  void publishSuccess() {
+    feedback_.success = true;
+    result_.success = true;
+    as_.publishFeedback(feedback_);
+    as_.setSucceeded(result_);
+  }
+
+  void executeCB(const behavior_actions::Fourber2023GoalConstPtr &goal)
+  {
+    ros::spinOnce();
+
+    // case where elevater node has given info that we are in a zone we need to be safe
+    if (goal->safety_position == fourber_ns::SAFETY_HIGH || goal->safety_position == fourber_ns::SAFETY_INTAKE_LOW) {
```

Comment:
Might want to error check the safety_position is a valid enum and return failure if not.  
I mean, other code should never do that, but if we don't check, we'll never know